File: srcs/event_func.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   event_func.c                                       :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: becaraya <becaraya@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2019/08/28 09:18:54 by becaraya          #+#    #+#             */
/*   Updated: 2019/12/12 23:23:55 by becaraya         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "doom_nukem.h"

static void	z_func(t_al *al)
{
	if (al->ev.type == SDL_KEYDOWN)
	{
		al->play.size = PLAYER_CROUCH;
		al->play.eyez -= PLAYER_SIZE - PLAYER_CROUCH;
		al->play.power_mult = 0.3;
	}
	else
	{
		al->play.size = PLAYER_SIZE;
		al->play.eyez += PLAYER_SIZE - PLAYER_CROUCH;
		al->play.power_mult = 1;
	}
}

static void	keyup_func(t_al *al, SDL_KeyboardEvent kev)
{
	kev.keysym.sym == SDLK_w ? al->k.w = 0 : 0;
	kev.keysym.sym == SDLK_a ? al->k.a = 0 : 0;
	kev.keysym.sym == SDLK_s ? al->k.s = 0 : 0;
	kev.keysym.sym == SDLK_d ? al->k.d = 0 : 0;
	kev.keysym.sym == SDLK_LEFT ? al->k.left = 0 : 0;
	kev.keysym.sym == SDLK_RIGHT ? al->k.righ = 0 : 0;
	kev.keysym.sym == SDLK_LSHIFT ? al->play.power = PLAYER_AERO_POWER : 0;
	kev.keysym.sym == SDLK_UP ? al->k.up = 0 : 0;
	kev.keysym.sym == SDLK_DOWN ? al->k.down = 0 : 0;
	kev.keysym.sym == SDLK_SPACE ? al->k.space = 0 : 0;
}

void		key_func(t_al *al)
{
	SDL_KeyboardEvent kev;

	kev = al->ev.key;
	kev.keysym.sym == SDLK_z ? z_func(al) : 0;
	if (al->ev.type == SDL_KEYDOWN)
	{
		kev.keysym.sym == SDLK_ESCAPE ? yeet(al) : 0;
		if (kev.keysym.sym == SDLK_F1)
			al->status = (al->status == GAME) ? PAUSE : GAME;
		kev.keysym.sym == SDLK_w ? al->k.w = 1 : 0;
		kev.keysym.sym == SDLK_a ? al->k.a = 1 : 0;
		kev.keysym.sym == SDLK_s ? al->k.s = 1 : 0;
		kev.keysym.sym == SDLK_d ? al->k.d = 1 : 0;
		kev.keysym.sym == SDLK_LEFT ? al->k.left = 1 : 0;
		kev.keysym.sym == SDLK_RIGHT ? al->k.righ = 1 : 0;
		kev.keysym.sym == SDLK_UP ? al->k.up = 1 : 0;
		kev.keysym.sym == SDLK_DOWN ? al->k.down = 1 : 0;
		kev.keysym.sym == SDLK_LSHIFT ? al->play.power = PLAYER_ANA_POWER : 0;
		kev.keysym.sym == SDLK_SPACE ? al->k.space = 1 : 0;
		kev.keysym.sym == SDLK_KP_MINUS ? al->fov *= 1.1 : 0;
		kev.keysym.sym == SDLK_KP_PLUS ? al->fov /= 1.1 : 0;
	}
	else
		keyup_func(al, kev);
}

File: web version/src/event_func.js
// Fonction pour gérer l'action de la touche "Z" (accroupissement)
function zFunc(al, isKeyDown) {
	if (isKeyDown) {
	  al.play.size = PLAYER_CROUCH;
	  al.play.eyez -= PLAYER_SIZE - PLAYER_CROUCH;
	  al.play.power_mult = 0.3;
	} else {
	  al.play.size = PLAYER_SIZE;
	  al.play.eyez += PLAYER_SIZE - PLAYER_CROUCH;
	  al.play.power_mult = 1;
	}
  }
  
  // Fonction pour gérer les touches relâchées
  function keyUpFunc(al, key) {
	switch (key) {
	  case "w": al.k.w = 0; break;
	  case "a": al.k.a = 0; break;
	  case "s": al.k.s = 0; break;
	  case "d": al.k.d = 0; break;
	  case "t": al.k.t = 0; break;
	  case "m": al.k.m = 0; break;
	  case "f": al.play.fly = 0; break;
	  case "[": al.k.lbrack = 0; break;
	  case "]": al.k.rbrack = 0; break;
	  case "ArrowLeft": al.k.left = 0; break;
	  case "ArrowRight": al.k.righ = 0; break;
	  case "Shift": al.play.power = PLAYER_AERO_POWER; break;
	  case "ArrowUp": al.k.up = 0; break;
	  case "ArrowDown": al.k.down = 0; break;
	  case " ": al.k.space = 0; break;
	}
  }
  
  // Fonction pour gérer les touches appuyées
  function keyDownFunc(al, key) {
	switch (key) {
	  case "Escape": prepareExit(al); break;
	  case "F1": al.status = (al.status === GAME) ? PAUSE : GAME; break;
	  case "w": al.k.w = 1; break;
	  case "a": al.k.a = 1; break;
	  case "s": al.k.s = 1; break;
	  case "d": al.k.d = 1; break;
	  case "t": al.k.t = 1; break;
	  case "m": al.k.m = 1; break;
	  case "f": al.play.fly = 1; break;
	  case "[": al.k.lbrack = 1; break;
	  case "]": al.k.rbrack = 1; break;
	  case "ArrowLeft": al.k.left = 1; break;
	  case "ArrowRight": al.k.righ = 1; break;
	  case "ArrowUp": al.k.up = 1; break;
	  case "ArrowDown": al.k.down = 1; break;
	  case "Shift": al.play.power = PLAYER_ANA_POWER; break;
	  case " ": al.k.space = 1; break;
	  case "-": al.fov *= 1.1; break;
	  case "+": al.fov /= 1.1; break;
	}
  }
  
  // Fonction principale pour gérer les événements clavier
  function keyFunc(al, event) {
	const key = event.key;
  
	if (key === "z") {
	  zFunc(al, event.type === "keydown");
	}
  
	if (event.type === "keydown") {
	  keyDownFunc(al, key);
	} else if (event.type === "keyup") {
	  keyUpFunc(al, key);
	}
  }